tests/mistral_utils.py

```diff
@@ -0,0 +1,267 @@
+        f"{i}.head_attention_value_output",
+        head_attention_value_output,
+        INTERVENTION_ACTIVATIONS,
+    )
+    CACHE_ACTIVATIONS[f"{i}.head_attention_value_output"] = head_attention_value_output
+    
+    attn_value_output = head_attention_value_output.transpose(1, 2).contiguous()
+    attn_value_output = attn_value_output.reshape(b, s, self_attn.hidden_size)
+
+    attn_value_output = DO_MISTRAL_INTERVENTION(
+        f"{i}.attention_value_output", attn_value_output, INTERVENTION_ACTIVATIONS
+    )
+    CACHE_ACTIVATIONS[f"{i}.attention_value_output"] = attn_value_output
+    attn_output = self_attn.o_proj(attn_value_output)
+    return attn_output
+
+
+def MISTRAL_MLP_RUN(mlp, hidden_states, i, CACHE_ACTIVATIONS, INTERVENTION_ACTIVATIONS):
+    hidden_states_c_fc = mlp.gate_proj(hidden_states)
+    hidden_states_act = mlp.act_fn(hidden_states_c_fc)
+
+    hidden_states_act = DO_MISTRAL_INTERVENTION(
+        f"{i}.mlp_activation", hidden_states_act, INTERVENTION_ACTIVATIONS
+    )
+    CACHE_ACTIVATIONS[f"{i}.mlp_activation"] = hidden_states_act
+    
+    hidden_states_c_proj = mlp.down_proj(hidden_states_act * mlp.up_proj(hidden_states))
+    return hidden_states_c_proj
+
+
+def MISTRAL_BLOCK_RUN(
+    block, hidden_states, position_ids, attention_mask, i, CACHE_ACTIVATIONS, INTERVENTION_ACTIVATIONS
+):
+    # self attention + residual
+    residual = hidden_states
+    hidden_states = block.input_layernorm(hidden_states)
+
+    hidden_states = DO_MISTRAL_INTERVENTION(
+        f"{i}.attention_input", hidden_states, INTERVENTION_ACTIVATIONS
+    )
+    CACHE_ACTIVATIONS[f"{i}.attention_input"] = hidden_states
+
+    attn_outputs = MISTRAL_SELF_ATTENTION_RUN(
+        block.self_attn, hidden_states, attention_mask, position_ids, i, CACHE_ACTIVATIONS, INTERVENTION_ACTIVATIONS
+    )
+
+    attn_outputs = DO_MISTRAL_INTERVENTION(
+        f"{i}.attention_output", attn_outputs, INTERVENTION_ACTIVATIONS
+    )
+    CACHE_ACTIVATIONS[f"{i}.attention_output"] = attn_outputs
+
+    attn_output = attn_outputs
+    # residual connection
+    hidden_states = attn_output + residual
+
+    # mlp + residual
+    residual = hidden_states
+    hidden_states = block.post_attention_layernorm(hidden_states)
+
+    hidden_states = DO_MISTRAL_INTERVENTION(
+        f"{i}.mlp_input", hidden_states, INTERVENTION_ACTIVATIONS
+    )
+    CACHE_ACTIVATIONS[f"{i}.mlp_input"] = hidden_states
+
+    feed_forward_hidden_states = MISTRAL_MLP_RUN(
+        block.mlp, hidden_states, i, CACHE_ACTIVATIONS, INTERVENTION_ACTIVATIONS
+    )
+
+    feed_forward_hidden_states = DO_MISTRAL_INTERVENTION(
+        f"{i}.mlp_output", feed_forward_hidden_states, INTERVENTION_ACTIVATIONS
+    )
+    CACHE_ACTIVATIONS[f"{i}.mlp_output"] = feed_forward_hidden_states
+
+    # residual connection
+    hidden_states = residual + feed_forward_hidden_states
+
+    return hidden_states
+
+
+def MISTRAL_RUN(mistral, input_ids, CACHE_ACTIVATIONS, INTERVENTION_ACTIVATIONS):
```

Comment:
nice! hopefully this can be used for mixtral as well (copy paste and MoE adaption is fine)